onnxruntime/core/providers/cuda/math/matmul.cc

```diff
@@ -88,6 +95,366 @@ static bool CanUseStridedBatchedGemm(const TensorShape& left_shape, const Tensor
+  // cublasLtMatmul only works if A is transposed and B is not.
+  cublasOperation_t transA = CUBLAS_OP_T;
+  cublasOperation_t transB = CUBLAS_OP_N;
+
+  const int M = static_cast<int>(helper.M());
+  const int K = static_cast<int>(helper.K());
+  const int N = static_cast<int>(helper.N());
+
+  const int lda = helper.Ldb(transA); // right_X transposed is N,K in row_major, so leading dimension is K in col_major
+  const int ldb = helper.Lda(transB); // left_X is M,K in row_major, so leading dimension is K in col_major
+  const int ldc = N; // result in col major (cublas) has the N,M shape
+  const int ldy = ldc;
+
+  // Create new tensors on the device which hold fp8 (__nv_fp8_e4m3) equivalents of left_X and right_X.
+  AllocatorPtr allocator;
+  ORT_RETURN_IF_ERROR(ctx->GetTempSpaceAllocator(&allocator));
+
+  size_t C_size = M * N * sizeof(MLFloat16);
+  IAllocatorUniquePtr<void> C = IAllocator::MakeUniquePtr<void>(allocator, C_size, false, ctx->GetComputeStream());
+
+  auto left_X_fp8 = IAllocator::MakeUniquePtr<void>(allocator, M * K * sizeof(Float8E4M3FN), false, ctx->GetComputeStream());
+  auto status = MLFloat16ToFloat8E4M3FN(stream, left_X, left_X_fp8.get());
+  if (! status.IsOK())
+    return status;
+
+  // If right_X is nullptr, it means that its value was constant and this computation was already done inside PrePack.
+  if (right_X != nullptr) {
+    auto status = TransposeAndConvertToFp8(stream, right_X, right_X_fp8, device_prop, allocator, ctx->GetComputeStream());
+    if (! status.IsOK()) {
+      return status;
+    }
+  }
+
+  // `cublasltmatmul` computes the following formula: D = alpha*(A*B) + beta*(C).
+  // Our matrix multiplication doesn't use a beta * C bias, so we just set the beta to 0 when calling the API.
+  // https://docs.nvidia.com/cuda/cublas/index.html?highlight=cublasLtMatmul#cublasltmatmul
+  float beta = 0;
+
+  const void* p_input_a = right_X_fp8.get();
+  const void* p_input_b = left_X_fp8.get();
+  const void* p_input_c = C.get();
+  void* p_output_y = Y->MutableDataRaw();
+
+  // Create matrix descriptors. Not setting any extra attributes.
+  // ORT is row_major, but cublas is col_major, so we need to use right_X as A and left_X as B.
+  int32_t dtype_A = right_X->GetElementType();
+  int32_t dtype_B = left_X->GetElementType();
+  int32_t dtype_C = Y->GetElementType();
+  int32_t dtype_Y = Y->GetElementType();
+  cudaDataType_t a_cuda_type = onnxruntime::cuda::ToCudaDataType(dtype_A);
+  cudaDataType_t b_cuda_type = onnxruntime::cuda::ToCudaDataType(dtype_B);
+  cudaDataType_t c_cuda_type = onnxruntime::cuda::ToCudaDataType(dtype_C);
+  cudaDataType_t y_cuda_type = onnxruntime::cuda::ToCudaDataType(dtype_Y);
+
+  // https://docs.nvidia.com/cuda/cublas/index.html?highlight=cublasStatus_t#cublasltmatmul
+  // "The compute type must be CUBLAS_COMPUTE_32F."
+  // "The scale type must be CUDA_R_32F."
+  cudaDataType_t scale_cuda_type = CUDA_R_32F;
```

Comment:
Please refer to https://docs.nvidia.com/cuda/cublas/index.html#id105 for FP8 GEMM.